### PR TITLE
ESQL: Fix memory tests

### LINKED ISSUE #103789: [CI] FilterOperatorTests testSimpleCircuitBreakingAtLimit failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/yd2ap6hcntx6g/tests/:x-pack:plugin:esql:compute:test/org.elasticsearch.compute.operator.FilterOperatorTests/testSimpleCircuitBreakingAtLimit

**Reproduction line:**
```
gradlew ':x-pack:plugin:esql:compute:test' --tests "org.elasticsearch.compute.operator.FilterOperatorTests.testSimpleCircuitBreakingAtLimit" -Dtests.seed=51E59C31506BA3C6 -Dtests.locale=zh-Hans-SG -Dtests.timezone=America/Belem -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
[Failure dashboard for `org.elasticsearch.compute.operator.FilterOperatorTests#testSimpleCircuitBreakingAtLimit`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testSimpleCircuitBreakingAtLimit'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.compute.operator.FilterOperatorTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure excerpt:**
```
junit.framework.AssertionFailedError: Expected exception CircuitBreakingException but no exception was thrown

  at __randomizedtesting.SeedInfo.seed([51E59C31506BA3C6:E9401855CDEA0F3D]:0)
  at org.apache.lucene.tests.util.LuceneTestCase.expectThrows(LuceneTestCase.java:2889)
  at org.apache.lucene.tests.util.LuceneTestCase.expectThrows(LuceneTestCase.java:2875)
  at org.elasticsearch.compute.operator.OperatorTestCase.testSimpleCircuitBreaking(OperatorTestCase.java:123)
  at org.elasticsearch.compute.operator.OperatorTestCase.testSimpleCircuitBreakingAtLimit(OperatorTestCase.java:105)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

### PR DESCRIPTION
**Pull request by nik9000**
This improves the memory tests of ESQL by removing the hard memory limits, instead calculating "how much memory is too much" on the fly. We do a binary search on the operation itself, looking for the first value that'll throw.

Closes #103789

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java [hunk 1]
```diff
@@ -73,69 +73,58 @@ public final void testSimpleLargeInput() {
...
-        testSimpleCircuitBreaking(memoryLimitForSimple());
+    public final void testSimpleCircuitBreaking() {
+        ByteSizeValue memoryLimitForSimple = enoughMemoryForSimple();
+        Operator.OperatorFactory simple = simple();
+        DriverContext inputFactoryContext = driverContext();
+        List<Page> input = CannedSourceOperator.collectPages(simpleInput(inputFactoryContext.blockFactory(), between(1_000, 10_000)));
+        try {
+            ByteSizeValue limit = BreakerTestUtil.findBreakerLimit(
+                memoryLimitForSimple,
+                l -> runWithLimit(simple, CannedSourceOperator.deepCopyOf(input), l)
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java:96]: This means we need to execute runWithLimit approximately 15 times to determine the lower bound if it's less than 100KB. Should we consider reducing the upper bound? However, I am fine with leaving it as is if there's no significant increase in time for CI.

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks, Nik!

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java:96]: It doesn't seem to mind too much. On my laptop running it for all 49 existing tests takes 8.6 seconds. That seems ok.

Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java:96]: Here's the logging of the slowest one:
```
[2024-01-08T11:43:05,418][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 1gb. 1gb bytes remaining.
[2024-01-08T11:43:05,473][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 512mb. 512mb bytes remaining.
[2024-01-08T11:43:05,528][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 256mb. 256mb bytes remaining.
[2024-01-08T11:43:05,581][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 128mb. 128mb bytes remaining.
[2024-01-08T11:43:05,637][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 64mb. 64mb bytes remaining.
[2024-01-08T11:43:05,692][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 32mb. 32mb bytes remaining.
[2024-01-08T11:43:05,696][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 16mb and 32mb. 16mb bytes remaining.
[2024-01-08T11:43:05,701][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 24mb and 32mb. 8mb bytes remaining.
[2024-01-08T11:43:05,759][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 24mb and 28mb. 4mb bytes remaining.
[2024-01-08T11:43:05,815][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 24mb and 26mb. 2mb bytes remaining.
[2024-01-08T11:43:05,820][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25mb and 26mb. 1mb bytes remaining.
[2024-01-08T11:43:05,825][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 26mb. 512kb bytes remaining.
[2024-01-08T11:43:05,880][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.7mb. 256kb bytes remaining.
[2024-01-08T11:43:05,937][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.6mb. 128kb bytes remaining.
[2024-01-08T11:43:05,996][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 64kb bytes remaining.
[2024-01-08T11:43:06,049][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 32kb bytes remaining.
[2024-01-08T11:43:06,104][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 16kb bytes remaining.
[2024-01-08T11:43:06,109][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 8kb bytes remaining.
[2024-01-08T11:43:06,114][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 4kb bytes remaining.
[2024-01-08T11:43:06,172][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 2kb bytes remaining.
[2024-01-08T11:43:06,225][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 1kb bytes remaining.
[2024-01-08T11:43:06,231][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 512b bytes remaining.
[2024-01-08T11:43:06,284][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 256b bytes remaining.
[2024-01-08T11:43:06,338][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 128b bytes remaining.
[2024-01-08T11:43:06,393][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 64b bytes remaining.
[2024-01-08T11:43:06,398][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 32b bytes remaining.
[2024-01-08T11:43:06,403][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 16b bytes remaining.
[2024-01-08T11:43:06,408][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 8b bytes remaining.
[2024-01-08T11:43:06,414][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 4b bytes remaining.
[2024-01-08T11:43:06,420][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 2b bytes remaining.
[2024-01-08T11:43:06,485][INFO ][o.e.c.a.CountDistinctLongGroupingAggregatorFunctionTests] [testSimpleCircuitBreaking] testing with 3.3mb against a limit of 25.5mb
[2024-01-08T11:43:06,487][INFO ][o.e.c.a.CountDistinctLongGroupingAggregatorFunctionTests] [testSimpleCircuitBreaking] after test
```

Seems ok to me.

Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java:96]: Yes, let's leave it as is.

**Pull request merged**